src/libraries/Common/src/Interop/Android/System.Security.Cryptography.Native.Android/Interop.Dsa.cs

```diff
@@ -0,0 +1,190 @@
+        }
+
+        [DllImport(Libraries.CryptoNative, EntryPoint = "AndroidCryptoNative_DsaSignatureFieldSize")]
+        private static extern int AndroidCryptoNative_DsaSignatureFieldSize(SafeDsaHandle dsa);
+
+        /// <summary>
+        /// Return the size of the 'r' or 's' signature fields in bytes.
+        /// </summary>
+        internal static int DsaSignatureFieldSize(SafeDsaHandle dsa)
+        {
+            // Add another byte for the leading zero byte.
+            int size = AndroidCryptoNative_DsaSignatureFieldSize(dsa);
+            Debug.Assert(size * 2 < DsaEncodedSignatureSize(dsa));
+            return size;
+        }
+
+        [DllImport(Libraries.CryptoNative, EntryPoint = "AndroidCryptoNative_DsaSizeP")]
+        private static extern int DsaSizeP(SafeDsaHandle dsa);
+
+        /// <summary>
+        /// Return the size of the key in bytes.
+        /// </summary>
+        internal static int DsaKeySize(SafeDsaHandle dsa)
+        {
+            int keySize = DsaSizeP(dsa);
+
+            // Assume an even multiple of 8 bytes \ 64 bits
+            keySize = (keySize + 7) / 8 * 8;
+            return keySize;
+        }
+
+        internal static bool DsaSign(SafeDsaHandle dsa, ReadOnlySpan<byte> hash, Span<byte> refSignature, out int outSignatureLength) =>
+            DsaSign(dsa, ref MemoryMarshal.GetReference(hash), hash.Length, ref MemoryMarshal.GetReference(refSignature), out outSignatureLength);
+
+        [DllImport(Libraries.CryptoNative, EntryPoint = "AndroidCryptoNative_DsaSign")]
+        [return: MarshalAs(UnmanagedType.Bool)]
+        private static extern bool DsaSign(SafeDsaHandle dsa, ref byte hash, int hashLength, ref byte refSignature, out int outSignatureLength);
+
+        internal static bool DsaVerify(SafeDsaHandle dsa, ReadOnlySpan<byte> hash, ReadOnlySpan<byte> signature)
+        {
+            int ret = DsaVerify(
+                dsa,
+                ref MemoryMarshal.GetReference(hash),
+                hash.Length,
+                ref MemoryMarshal.GetReference(signature),
+                signature.Length);
+
+            if (ret == -1)
+            {
+                throw new CryptographicException();
+            }
+
+            return ret == 1;
+        }
+
+        [DllImport(Libraries.CryptoNative, EntryPoint = "AndroidCryptoNative_DsaVerify")]
+        private static extern int DsaVerify(SafeDsaHandle dsa, ref byte hash, int hashLength, ref byte signature, int signatureLength);
+
+        internal static DSAParameters ExportDsaParameters(SafeDsaHandle key, bool includePrivateParameters)
+        {
+            Debug.Assert(
+                key != null && !key.IsInvalid,
+                "Callers should check the key is invalid and throw an exception with a message");
+
+            if (key == null || key.IsInvalid)
+            {
+                throw new CryptographicException();
+            }
+
+            IntPtr p_bn, q_bn, g_bn, y_bn, x_bn; // these are not owned
```

Comment:
Aren't they?  The GetDsaParameters function does (e.g.) `*p = ToGRef(stuff)`, so doesn't the managed code now need to call FreeGlobalRef on it?  Seems like these should be SafeSomethingHandles.

Possibly this got missed for other algorithm types.

Reply:
Yep. I missed that OpenSSL didn't hand off ownership. I'll update the interop methods for all the algorithms I've done to fix ownership and remove these leaks.